compiler/rustc_type_ir/src/lib.rs

```diff
@@ -359,3 +359,66 @@ rustc_index::newtype_index! {
     #[gate_rustc_only]
     pub struct BoundVar {}
 }
+
+/// Represents the various closure traits in the language. This
+/// will determine the type of the environment (`self`, in the
+/// desugaring) argument that the closure expects.
+///
+/// You can get the environment type of a closure using
+/// `tcx.closure_env_ty()`.
+#[derive(Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Hash, Debug)]
+#[cfg_attr(feature = "nightly", derive(Encodable, Decodable, HashStable_NoContext))]
+pub enum ClosureKind {
+    // Warning: Ordering is significant here! The ordering is chosen
+    // because the trait Fn is a subtrait of FnMut and so in turn, and
+    // hence we order it so that Fn < FnMut < FnOnce.
+    Fn,
+    FnMut,
+    FnOnce,
+}
+
+impl ClosureKind {
+    /// This is the initial value used when doing upvar inference.
+    pub const LATTICE_BOTTOM: ClosureKind = ClosureKind::Fn;
+
+    pub const fn as_str(self) -> &'static str {
+        match self {
+            ClosureKind::Fn => "Fn",
+            ClosureKind::FnMut => "FnMut",
+            ClosureKind::FnOnce => "FnOnce",
+        }
+    }
+
+    /// Returns `true` if a type that impls this closure kind
+    /// must also implement `other`.
+    pub fn extends(self, other: ClosureKind) -> bool {
+        self <= other
+    }
+
+    /// Converts `self` to a `DefId` of the corresponding trait.
+    ///
+    /// Note: the inverse of this function is `TyCtxt::fn_trait_kind_from_def_id`.
+    pub fn to_def_id<I: Interner>(self, interner: I) -> I::DefId {
+        match self {
+            ClosureKind::Fn => interner.fn_def_id(),
+            ClosureKind::FnMut => interner.fn_mut_def_id(),
+            ClosureKind::FnOnce => interner.fn_once_def_id(),
+        }
+    }
+
+    /// Returns the representative scalar type for this closure kind.
+    /// See `Ty::to_opt_closure_kind` for more details.
+    pub fn to_ty<I: Interner>(self, interner: I) -> I::Ty {
+        match self {
+            ClosureKind::Fn => interner.i8_type(),
+            ClosureKind::FnMut => interner.i16_type(),
+            ClosureKind::FnOnce => interner.i32_type(),
+        }
+    }
```

Comment:
This seems better as methods on `Interner` itself? Rather than `rustc_ty_ir` knowing about this directly.

Reply:
That would require changing every call site in the codebase that currently uses these functions though. I also think we're really moving towards *fewer* functions on tcx, not more 😅

Reply:
> So, I don't think the "fewer functions" argument holds up well.

I mean that we'd need to add it as an inherent method to `TyCtxt<'tcx>`, and not a trait method of the `Interner`, which most non-`rustc_type_ir` code doesn't import (and therefore which doesn't pollute the type namespace).

`Interner` can have as many functions as needed, and it probably will end up with a lot of weird shims because of the uneven API surface it's trying to capture.

> I also could imagine that we don't want to bake-in the idea that closure kinds must be repesentable as types, anyways.

Yeah, though rustc needs it, and unfortunately we can't define inherent functions on `ClosureKind` in `rustc_middle` that only apply to the `Interner = TyCtxt<'tcx>` case.

---

I think I will just pull these into `TyCtxt<'tcx>` functions (not `Interner`) though, after second though.